### PR TITLE
Perfomance metrics tutorial

### PR DESCRIPTION
**Pull request by ahcorde**
This PR is related with this other PR https://github.com/osrf/gazebo/pull/2819 which includes a new message called `performanceMetrics.proto`. This tutorial it's to understand the readings.

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

File: performance_metrics/tutorial.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
+# Performance metrics
+
+Gazebo 9 and Gazebo 11 publish a message called `/gazebo/performance_metrics` that allows to check the performance
+of each sensor in the world.
+
+# Learn how to read the measures
```
Comment (iche033) [inline performance_metrics/tutorial.md:6]: `measurements`?

File: performance_metrics/tutorial.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  sensor_name: "default::camera_model::link::camera_sensor_regular"
+  real_sensor_update_rate: 0.29411702534733497
+  sim_sensor_update_rate: 29.411764705882351
+  fps: 2.0637693405151367
+}
+```
+
+#### Run the world without lockstep
+
+Again in this case, where we run the world without `lockstep`, there is no guarantee when the callback is called, and the `sim_sensor_update_rate` value is highly impacted, it's should be close to 500 but in this case it's close to 0.
```
Comment (iche033) [inline performance_metrics/tutorial.md:86]: `it's should be` -> `it should be`

File: performance_metrics/tutorial.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+- real_sensor_update_rate equals 100
+[[file:files/contact_100.png|600px]]
+- real_sensor_update_rate equals 10
+[[file:files/contact_10.png|600px]]
+
+In this case, this type of sensor are not blocking the main thread which means that `Server::Run` will use always the same time (~100ms). This means that we are going to miss some reading.
+
+### Run the world without lockstep
+
+If `lockstep` is disabled then `Server::Run` will run as fast as possible, we are not going to miss any reading, we will see always a `sim_sensor_update_rate` closed to 1000 even if we change the `real_sensor_update_rate`.
```
Comment (iche033) [inline performance_metrics/tutorial.md:134]: `closed to` -> `close to`

File: performance_metrics/tutorial.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+sensor {
+  sensor_name: "default::model_1::link_1::sensor_contact"
+  real_sensor_update_rate: 900.55041641451237
+  sim_sensor_update_rate: 1000
+}
+```
+
+### Run the world with lockstep
+
+If `lockstep` is enabled the `Server::Run` will run at 100ms if there is no a rendering-based sensor in the world that may change this value as explained before.
```
Comment (iche033) [inline performance_metrics/tutorial.md:147]: `.. at 100ms if there is no a..` -> `..at 100ms. If there is not a..`?

**Comment by ahcorde** [comment]
@iche033 I updated the tutorial with the changes made on the performance metrics PR

File: performance_metrics/tutorial.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+sensor {
+  sensor_name: "default::camera_model::link::camera_sensor"
+  real_sensor_update_rate: 96.478915160204693
+  sim_sensor_update_rate: 0.14425851125216388
+  fps: 80.7220458984375
+}
+sensor {
+  sensor_name: "default::camera_model::link::camera_sensor_regular"
+  real_sensor_update_rate: 169.40730989154036
+  sim_sensor_update_rate: 0.14452955629426217
```
Comment (iche033) [inline performance_metrics/tutorial.md:100]: interesting. If we slow down physics by changing the `real_time_update_rate` to `10`, the camera sensor should have more time to catch up and hit it's target rate

I'm seeing quite different values from the ones here using the latest changes that compute metrics inside SensorManager.

The `real_sensor_update_rate` for me is close to ~0.3, while `sim_sensor_update_rate` is ~30. Same for the high speed camera above, I'm getting close to ~5 `real_sensor_update_rate` and ~500 `sim_sensor_update_rate`

File: performance_metrics/tutorial.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+The maximum time step size that can be taken by a variable time-step solver (such as simbody) during simulation.
+For physics engines with fixed-step solvers (like ODE), this is simply the time step size. The default value
+in Gazebo is 0.001 seconds.
+
+When you choose the `update rate` of a sensor you need to take in account if you have enough precision
+defined in `max step size`. For example:
+
+  - 1/250 = 0.004 it's fine using the default value.
+  - 1/400 = 0.0025. You will need to increase the `max step size` to 0.0001, this will make the simulation slower.
```
Comment (iche033) [inline performance_metrics/tutorial.md:115]: `increase` -> `reduce`?

Can you mention that reducing max step size also has side effect on the accuracy and speed of physics simulation, and you can point them to this tutorial which gives good description of what the params do:
http://gazebosim.org/tutorials?tut=physics_params&cat=physics

File: performance_metrics/tutorial.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+We are going to compare the results between physics (contact) and rendering (camera) based sensors and the differences
+when the `lockstep` flag is active.
+
+## Camera
+
+For the rendering-based sensor we are going to use the [camera_strict_rate.world](https://github.com/osrf/gazebo/blob/gazebo9/test/worlds/camera_strict_rate.world) world. It contains two cameras: a high-resolution (1280x720) high-frame-rate (500 fps) and a a low-resolution (320x240) low-frame-rate (30 fps) camera.
+
+### Run the world without lockstep
+
+With the `lockstep` flag deactivated there is no guarantee when the callback is called. This means the `sim_sensor_update_rate` could not correspond with the real value defined in the world. Inspect the `sim_sensor_update_rate` field, It is likely unable to reach the specified 500 fps. The other low-resolution camera is able to reach the specified value. Take note of the `real time factor` it should be close to 1.0.
```
Comment (iche033) [inline performance_metrics/tutorial.md:17]: may want to elaborate on what `callback` is, e.g.

```
..there is no guarantee that the camera sensor update happens in the same iteration as the physics simulation.
```

there is also another mention of `callback` below

File: performance_metrics/tutorial.md [hunk 8]
```diff
@@ -0,0 +1,142 @@
+# Performance metrics
+
+Gazebo 9 and Gazebo 11 publish a message called `/gazebo/performance_metrics` that allows to check the performance
+of each sensor in the world.
+
+# Learn how to read the measurements
+
+We are going to compare the results between physics (contact) and rendering (camera) based sensors and the differences
```
Comment (iche033) [inline performance_metrics/tutorial.md:8]: the tutorial no longer has the physics (contact) sensor values?

File: performance_metrics/tutorial.md (hunk 5 above)
Comment (ahcorde) [reply performance_metrics/tutorial.md:100]: Added instantaneous RTF in the PR : https://github.com/osrf/gazebo/pull/2819#discussion_r481971068

**Comment by ahcorde** [comment]
- Fixed real time factor in the Gazebo PR
 - Included contact sensor section
 - Link to Physics parameters

File: performance_metrics/tutorial.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+  real_sensor_update_rate: 1011.0170528246299
+  sim_sensor_update_rate: 1000
+}
+```
+
+### Run the world without lockstep
+
+Again in this case, where we run the world without `lockstep`, there is no guarantee that the camera sensor update happens
+in the same iteration as the physics simulation, and the `sim_sensor_update_rate` value is highly impacted, it should be
+close to 500 but in this case it's close to 0.
```
Comment (iche033) [inline performance_metrics/tutorial.md:131]: the description here describes update rates for camera sensors but the values below are from contact sensor

File: performance_metrics/tutorial.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+  real_sensor_update_rate: 1006.3885545442469
+  sim_sensor_update_rate: 1000
+}
+```
+
+## Real time update rate
+
+We can modify the Real time update rate in the Physics Engine. As you may see the `sim_sensor_update_rate` in both cameras
+are the expected value 500 and 30. In this case the Real time update rate is equal to **10** which make the simulation slower. The fps
+are lower, it's able to generate almost 6 frames for the high speed camera and 2 for the regular camera each *real second*.
```
Comment (iche033) [inline performance_metrics/tutorial.md:107]: hmm we're not showing metrics for camera sensors in this section, that probably got accidentally removed?

**Comment by iche033** [comment]
I made some changes in a3342cddf05750fe58106832befe13982bb2fe96. Most of them are minor. I removed the framerate numbers in the Real time update section as it does not correspond to the values shown and added an explanation for the values that are 0.

**Review by iche033 — APPROVED** [review]

**Pull request merged**